packages/oidc/src/token.ts

```diff
@@ -0,0 +1,226 @@
+interface TokenPayload {
+  sub: string;
+  name: string;
+  exp: number;
+}
+
+function isNodeEnv(): boolean {
+  return (
+    typeof process !== 'undefined' &&
+    process.versions != null &&
+    process.versions.node != null
+  );
+}
+
+function isBrowserEnv(): boolean {
+  return typeof window !== 'undefined';
+}
+
+async function getVercelDataDir(): Promise<string | null> {
+  if (isBrowserEnv()) {
+    return '';
+  }
+  const vercelFolder = 'com.vercel.cli';
+  const os = await import('os');
+  const path = await import('path');
+  if (process.env.XDG_DATA_HOME) {
+    return path.join(process.env.XDG_DATA_HOME, vercelFolder);
+  }
+  switch (os.platform()) {
+    case 'darwin':
+      return path.join('~/Library/Application Support', vercelFolder);
+    case 'linux':
+      return path.join('~/.local/share', vercelFolder);
+    case 'win32':
+      return path.join('%LOCALAPPDATA%', vercelFolder);
+    default:
+      return null;
+  }
+}
+
+async function getVercelCliToken(): Promise<string | null> {
+  const dataDir = await getVercelDataDir();
+  if (!dataDir) {
+    return null;
+  }
+  const path = await import('path');
+  const fs = await import('fs');
+  const tokenPath = path.join(dataDir, 'auth.json');
+  if (!fs.existsSync(tokenPath)) {
+    return null;
+  }
+  const token = fs.readFileSync(tokenPath, 'utf8');
+  return JSON.parse(token).token;
+}
+
+interface VercelTokenResponse {
+  token: string;
+}
+
+async function getVercelOidcToken(
+  authToken: string,
+  projectId: string
+): Promise<VercelTokenResponse | null> {
+  try {
+    const url = `https://api.vercel.com/projects/token?projectId=${projectId}&source=vercel-oidc-refresh`;
+    const res = await fetch(url, {
+      method: 'POST',
+      headers: {
+        Authorization: `Bearer ${authToken}`,
+      },
+    });
+    if (!res.ok) {
+      throw new Error(`failed to refresh token: ${res.statusText}`);
+    }
+    if (!res.json) {
+      throw new Error('failed to refresh token: no json response');
+    }
+    const tokenRes = (await res.json()) as VercelTokenResponse;
```

Comment:
WDYT about having a separate function for validating the response. We don't need Zod, we could just do something like

```ts
function assertVercelTokenResponse(input: unknown): asserts input is VercelTokenResponse {
  if (!input || typeof input !== 'object') {
    throw new TypeError('Expected an object');
  }
  if (!('token' in input) || typeof input.token !== 'string') {
    throw new TypeError('Expected a string-valued token property');
  }
  return input;
}
```